### ISSUE TITLE
Enhance test message when results are merged with `rebot --merge`

### ISSUE THREAD

**Issue by tysonite** [robotframework/robotframework#3319]
Let's say we have a suite with 3 test cases, on a 1st run last 2 fail, on 2nd run last one fails, on 3rd run all pass.

It is a bit confusing that A.C_1 (failed twice, passed once) test has 4 statuses:
```
Re-executed test has been merged.
New status: PASS
New message: 
Old status: FAIL
Old message: Re-executed test has been merged.
New status: FAIL
New message: AssertionError
Old status: FAIL
Old message: AssertionError
```
I use `rebot --merge output_1.xml output_2.xml output_3.xml` to merge 3 outputs.

Refer to logs: [logs.zip](https://github.com/robotframework/robotframework/files/3659270/logs.zip)

**Comment by pekkaklarck** [comment]
How'd you want it to look?

**Comment by tysonite** [comment]
I would say that I expect it to be something like that:
```
Re-executed test has been merged.
New status: PASS
New message: 
Old status: FAIL
Old message: AssertionError
Old status: FAIL
Old message: AssertionError
```

**Comment by pekkaklarck** [comment]
OK, I think I got it now. It seems output 1 is first merged with output 2 and that result is then merged with output 3. Possibly all merges could be done in one go and then the error message could be enhanced as well. I'm afraid implementing it would require quite big changes to the merge logic, though.

Although I can see this could be enhanced, I don't consider the current message that bad. Due to all other higher priority issues, I doubt I have time to look at this in the foreseeable future. Unless someone else is interested to look at this, or is willing to pay specially for fixing this, I will close this issue.

**Comment by yahman72** [comment]
The problem is just the confusing message right? In that case this could also be fixed without changes to the result merging logic.

This  could be done by updating just the message merging logic i.e. if the old message contains that "Re-executed test has been merged" string then modify the old message accordingly.
This should be quite simple - the files and methods to change would be more or less the same as in PR #2562  (where the message format was changed to HTML)

**Comment by pekkaklarck** [comment]
Yeah, I guess that would work as well. I'd don't like the idea of making guesses based on earlier messages that much, but probably it would be possible to add such markers when merging the first messages that subsequent merges could reliably detect that.

If this is done, it would also change the behavior in this case:
```
rebot --output merge1.xml --merge orig1.xml orig2.xml
rebot --output merge2.xml --merge merge1.xml orig3.xml
```
I guess that could be considered a feature rather than a bug.

**Comment by pekkaklarck** [comment]
Do you @tysonite still consider this a problem? Would enhancing the message be enough? That ought to be pretty easy to implement.

**Comment by tysonite** [comment]
I think it would be more user-friendly, if messages reflect more correctly the results in HTML log. But this is something really related to user experience rather than to issues. Even if not fixed, this could be explained to someone who looks into HTML logs.

**Comment by pekkaklarck** [comment]
Do you have concrete ideas what to do to this issue? Changing the message would be easy, easily done before RF 3.2 final, anything else much more complicated.

**Comment by tysonite** [comment]
I suppose merged messages shall look like below for the `rebot --merge output_1.xml output_2.xml output_3.xml` case:
```
Re-executed test has been merged.
New status: PASS
New message: 

Old status: FAIL
Old message: AssertionError

Old status: FAIL
Old message: AssertionError
```

Would it be possible with just message changing?

**Comment by tysonite** [comment]
I did some experiments with the file you mentioned earlier:
```
diff --git a/src/robot/result/merger.py b/src/robot/result/merger.py
index 8f1a46852..8127a2ced 100644
--- a/src/robot/result/merger.py
+++ b/src/robot/result/merger.py
@@ -87,13 +87,21 @@ class Merger(SuiteVisitor):
             return html_escape(message)

     def _create_merge_message(self, new, old):
-        return ''.join([
-            '*HTML* Re-executed test has been merged.<hr>',
-            'New status: %s<br>' % self._format_status(new.status),
-            'New message: %s<hr>' % self._html_escape(new.message),
-            'Old status: %s<br>' % self._format_status(old.status),
-            'Old message: %s' % self._html_escape(old.message)
-        ])
+        if 'Re-executed test has been merged' in old.message:
+            return ''.join([
+                '*HTML* Re-executed test has been merged.<hr>',
+                'New status: %s<br>' % self._format_status(new.status),
+                'New message: %s<hr>' % self._html_escape(new.message),
+                '%s' % self._html_escape(old.message.replace('New', 'Old').replace('Re-executed test has been merged.<hr>', ''))
+           ])
+        else:
+            return ''.join([
+                '*HTML* Re-executed test has been merged.<hr>',
+                'New status: %s<br>' % self._format_status(new.status),
+                'New message: %s<hr>' % self._html_escape(new.message),
+                'Old status: %s<br>' % self._format_status(old.status),
+                'Old message: %s' % self._html_escape(old.message)
+            ])

     def _format_status(self, status):
         return '<span class="%s">%s</span>' % (status.lower(), status)
```

It looks like that this patch does what you mean.

**Comment by pekkaklarck** [comment]
That message looks good to me and I guess this would be an ok approach in general, but just looking that the message contains something and especially unconditionally replacing "New" with "Old" are both pretty fragile. We definitely don't want to change someone message containing e.g. "New York" to "Old York". Using more explicit markers would be a lot safer. Perhaps something like this:

- When creating the first merge message, start it with something like `*HTML* <span class="merge">Re-executed tests have been merged.</span><hr>` (somehow 'tests have' sounds better to my ear than 'test has')
- Use something like `<span class="new-status">New status:</span>` and `<span class="old-message">Old message:</span>` instead of literal strings.
- When checking is the message already a merge message, check does it start with `*HTML* <span class="merge">....</span><hr>` i.e. the exact text we added ourselves. If it does, split the old header away by using something like `message = old.message.split('<hr>', 1)[1]`.
- When creating the new message, replace `<span class="new-status">` with `<span class="old-status">`. It's _very_ unlikely someone has had this in their real message.
- For bonus points, don't include old or new messages that are empty in the message.

There's quite a lot of different cases here and I would expect it to be best to test this thoroughly with unit tests. Just needed to create a helper method accepting old and new messages that would be easy to test on that level. The old acceptance tests should obviously be updated and possibly some new added as well.

If you think the above make sense, please create a PR and I promise to look at it. I'm just about to start my winter holiday, but latest after that.

**Comment by tysonite** [comment]
Thanks for suggestions and guidelines, I've tried to encaplusate all of them. BTW, that's was funny to do such minor change. I've submited PR with proposed changes, please, have a look. I am sure that I have not modified unit tests as you wished as I am not completely sure how to do that properly - probably, new test file has to be created as well as `Merger` API has to be modified for better testability. But I am not certain how you want it to look like. Please, suggest.

**Comment by pekkaklarck** [comment]
I had totally forgotten about this issue. Should have added it to RF 3.2 scope. On a quick look the PR looks so good that I guess this can still be included.

**Comment by pekkaklarck** [comment]
This was implemented in PR #3476 by @tysonite and I refactored the code a little in 16762151c75238f0743467fdc69ca73d927a22a7.
